File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/IfChainWithoutElse.vb
Imports System
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Linq
Imports System.Text

Namespace Tests.TestCases
    Public Class IfChainWithoutElse
        Public Sub New(a As Boolean)
            If a Then
            End If

            If a Then
            Else
            End If

            If a Then
            ElseIf a Then
            ElseIf a Then ' Noncompliant {{Add the missing 'Else' clause with either the appropriate action or a suitable comment as to why no action is taken.}}
'           ^^^^^^
            End If

            If a Then
            ElseIf a Then
            ElseIf a Then
            ElseIf a Then
            ElseIf a Then
            ElseIf a Then
            ElseIf a Then
            ElseIf a Then
            ElseIf a Then
            ElseIf a Then ' Noncompliant
            End If

            If a Then
            ElseIf a Then
            ElseIf a Then ' Noncompliant
            Else
            End If

            If a Then
            ElseIf a Then
            ElseIf a Then
            ElseIf a Then
            ElseIf a Then
            ElseIf a Then
            ElseIf a Then
            ElseIf a Then
            ElseIf a Then
            ElseIf a Then ' Noncompliant
            Else
            End If

            If a Then
            ElseIf a Then
            ElseIf a Then
            Else
                Console.WriteLine()
            End If

            If a Then
            ElseIf a Then
            ElseIf a Then
            Else
                ' Single line comment
            End If

            If a Then
            ElseIf a Then
            ElseIf a Then
            Else ' Single line comment
            End If

            If a Then
            ElseIf a Then
            ElseIf a Then ' Noncompliant
            Else
            End If ' Single line comment

            If a Then
            ElseIf a Then
            ElseIf a Then
            Else
#If DEBUG
                Trace.WriteLine("Something to log only in debug", a.ToString())
#End If
            End If

            If a Then
            ElseIf a Then
            ElseIf a Then ' Noncompliant
            Else
#If DEBUG
#End If
            End If

            If a Then
            Else
                If a Then
                    If a Then
                    ElseIf a Then ' Noncompliant
                    End If
                End If
            End If
        End Sub
    End Class
End Namespace

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/IfChainWithoutElse.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace Tests.Diagnostics
{
    public class IfChainWithoutElse
    {
        public IfChainWithoutElse(bool a)
        {
            if (a)
            {
            }

            if (a)
            {
            }
            else
            {
            }

            if (a)
            {
            }
            else if (a)
            {
            }
            else if (a) // Noncompliant {{Add the missing 'else' clause with either the appropriate action or a suitable comment as to why no action is taken.}}
//          ^^^^^^^
            {
            }


            if (a) { }
            else if (a) { }
            else if (a) { }
            else if (a) { }
            else if (a) { }
            else if (a) // Noncompliant {{Add the missing 'else' clause with either the appropriate action or a suitable comment as to why no action is taken.}}
//          ^^^^^^^
            {
            }

            if (a) { }
            else if (a) { }
            else if (a) { } // Noncompliant
            else { }

            if (a) { }
            else if (a) { }
            else if (a) { }
            else if (a) { }
            else if (a) { }
            else if (a) { }
            else if (a) { } // Noncompliant
            else { }

            if (a) { }
            else if (a) { }
            else if (a) { }
            else
            {
                Console.WriteLine();
            }

            if (a) { }
            else if (a) { }
            else if (a) { }
            else
            {
                // Single line comment
            }

            if (a) { }
            else if (a) { }
            else if (a) { }
            else
            {
                /* Multi line comment
                 * Which is actually multi line
                 */
            }

            if (a) { }
            else if (a) { }
            else if (a) { }
            else
                Console.WriteLine();

            if (a) { }
            else if (a) { }
            else if (a) { }
            else
            {
#if DEBUG
            Trace.WriteLine("Something to log only in debug", a.ToString());
#endif
            }

            if (a) { }
            else if (a) { }
            else if (a) { } // Noncompliant
            else
            {
#if DEBUG
#endif
            }

            if (a)
            {
            }
            else
            {
                if (a)
                {
                    if (a)
                    {
                    }
                    else if (a) // Noncompliant
                    {
                    }
                }
            }
        }
    }
}
